src/github.com/matrix-org/dendrite/appservice/routing/user.go

```diff
@@ -0,0 +1,64 @@
+package routing
+
+import (
+	"encoding/json"
+	"github.com/matrix-org/dendrite/clientapi/jsonerror"
+	"github.com/matrix-org/dendrite/common/config"
+	"github.com/matrix-org/util"
+	"io/ioutil"
+	"net/http"
+)
+
+// URIToUIDResponse represents response to an AppService URI to User Id
+// (/_matrix/app/r0/user?uri={url_encoded_uri}) request
+type URIToUIDResponse struct {
+	UserID string `json:"user_id"`
+}
+
+// URIToUID implements `/_matrix/app/r0/user?uri={url_encoded_uri}`, which
+// enables users to contact App Service users directly by taking an encoded
+// URI and turning it into a Matrix ID on the homeserver.
+// https://matrix.org/docs/spec/application_service/unstable.html#user-ids
+func URIToUID(req *http.Request, cfg config.Dendrite) util.JSONResponse {
+	uri := req.URL.Query().Get("uri")
+	if uri == "" {
+		return util.JSONResponse{
+			Code: http.StatusOK,
+			JSON: nil,
+		}
+	}
+	for _, appservice := range cfg.Derived.ApplicationServices {
```

Comment:
It would be ideal to contact all application services simultaneously in parallel and then collect and present the results once all have answered or timed out.

You can see an example of doing so here: https://github.com/matrix-org/dendrite/pull/561/files#diff-f3464f1c9dd37b2cdcc0d85ac980d65fR195

Reply:
That PR has a lot of helpful stuff in it! I may have to refactor mine after that one gets merged...